### PR TITLE
refactor: Replace boost::bind with std::bind

### PR DESCRIPTION
**Pull request by ken2812221**
Replace boost::bind with std::bind

- In `src/rpc/server.cpp`, replace `std::transform` with simple loop.
- In `src/validation.cpp`, store the `boost::signals2::connection` object and use it to disconnect.
- In `src/validationinterface.cpp`, use 2 map to store the `boost::signals2::scoped_connection` object.

### CODE REVIEW

**Comment by fanquake** [comment]
Concept ACK

After this `boost::bind` will still be used by server, scheduler and validation (and associated tests). @ken2812221 Is there a particular reason you haven't removed that usage?

**Comment by practicalswift** [comment]
Concept ACK, but please remove `using namespace std::placeholders;`. Reference `_N` using the fully specified `std::placeholders::_N`.

See developer notes:

> - Don't import anything into the global namespace (`using namespace ...`). Use fully specified types such as `std::string`.

**Comment by ken2812221** [comment]
>Is there a particular reason you haven't removed that usage?

Those instances cannot be replaced simply `boost` -> `std`. I'll try to get rid of those.

>please remove `using namespace std::placeholders;`

Will do

**Comment by MarcoFalke** [comment]
I believe it is acceptable to be using `using namespace std::placeholders` here, since the alternative would be extremely verbose.

**Comment by practicalswift** [comment]
@MarcoFalke I'm not sure I agree. The verbosity could help remind us that we should use lambdas instead of ugly `std::bind` (modulo the few places where that is not technically possible in C++11), no? :-)

**Comment by ken2812221** [comment]
@practicalswift Since @MarcoFalke think that `using namespace std::placeholders;` is acceptable,  I will wait for more comments from other people and decide what to do.

**Comment by ken2812221** [comment]
@practicalswift Good point! I'll try to use lambda to replace bind

File: src/validationinterface.cpp
Comment (Empact) [inline src/validationinterface.cpp]: Can `std::move` `connections`

File: src/validationinterface.cpp
Comment (Empact) [inline src/validationinterface.cpp]: nit: could extract these as a method on `ValidationInterfaceConnections` where it would be more clear whether the method was comprehensive in its disconnecting.

File: src/validationinterface.cpp
Comment (Empact) [inline src/validationinterface.cpp]: nit: could extract as a method on `ValidationInterfaceConnections`

**Comment by promag** [comment]
Concept ACK.

File: src/validationinterface.h
Comment (MarcoFalke) [inline src/validationinterface.h]: I believe this will slow down the build significantly. So tend to NACK here.

Comment (ken2812221) [reply src/validationinterface.h]: Fixed

**Comment by MarcoFalke** [comment]
Could you sum up the differences between `std::bind` and lambda in the first comment for the convenience of the reviewers? It might not be obvious to everyone why lambdas should be preferred over `std::bind` everywhere.

**Comment by practicalswift** [comment]
@MarcoFalke Stephan T. Lavavej (standard library maintainer Visual C++) sums up the `std::bind` issues/gotchas here: https://www.youtube.com/watch?v=zt7ThwVfap0&t=1721

His recommendations:

> * Avoid using `bind()`
> * Use lambdas, especially generic lambdas

**Comment by theuni** [comment]
Please break this up into logical commits for review.
boost::bind -> std::bind makes sense to me. Concept ACK there for sure.

std::bind -> lambda seems more like a style/preference change. I prefer lambdas too, but I don't think that justifies the churn here. Maybe when lambdas are more robust and provide a more clear win with c++14/c++17?

**Comment by fanquake** [comment]
Travis linter is failing with:
```
Running script for: fcba334f0d21161354f168a081d95a63b82fa4c5
ALL_FILES=$(git ls-files -- '*.cpp' '*.h')
for i in 1 2 3 4 5; do
    sed -i "s/ _${i}/ std::placeholders::_${i}/g" ${ALL_FILES}
done
fatal: ambiguous argument '5': unknown revision or path not in the working tree.
```

**Comment by ken2812221** [comment]
Update: In this PR, I do boost::bind to std::bind only because there are some disputes about lambda.

**Comment by practicalswift** [comment]
@ken2812221 Makes sense! Getting rid of `boost/bind.hpp` is a very welcome first step :-)

**Comment by Empact** [comment]
Concept ACK, the callers should include `<functional>`

File: src/validation.cpp
Comment (ryanofsky) [inline src/validation.cpp]: Could drop this disconnect line (and the whole destructor) by replacing ` boost::signals2::connection` with `boost::signals2::scoped_connection` above.

File: src/validationinterface.cpp
Comment (ryanofsky) [inline src/validationinterface.cpp]: It seems like these two maps would most naturally be `MainSignalsInstance` members. I don't think it is good to arbitrary divide the validationinterface state up between `g_signals->m_internals` and these new global variables (unless there's a specific reason for doing this that I'm not seeing).

File: src/validationinterface.cpp
Comment (ryanofsky) [inline src/validationinterface.cpp]: I think if you used `scoped_connection` instead of `connection` types you could get rid of this for loop.

Also, since this for loop is the only thing that appears to iterate over `std::array<9>` objects, it might be a good idea to replace `std::array` with a plain struct if you do eliminate this loop. (I noticed you used a `ValidationInterfaceConnections` struct in a previous version of this PR).

**Review by ryanofsky — APPROVED** [review]
utACK d4a4f5729f0faa3bc51c6f162a10812c9703d5a5. I also think it'd be good to switch to lambdas in the future.

File: src/validationinterface.cpp
Comment (ken2812221) [reply src/validationinterface.cpp]: I was trying to do this however MainSignalsInstance is not created when `RegisterWithMempoolSignals()` called. So I only move `g_connMainSignals` into a member of `MainSignalsInstance`. Keeping `g_connNotifyEntryRemoved` as a static global variable.

File: src/validationinterface.cpp
Comment (ken2812221) [reply src/validationinterface.cpp]: That idea looks good to me. I'll take it.

**Comment by sipa** [comment]
Can this be implemented as a scripted diff? It looks relatively straightforward.

**Comment by ken2812221** [comment]
@sipa Done. Apparently I can't use i as the script's variable.

File: src/validationinterface.cpp
Comment (ryanofsky) [inline src/validationinterface.cpp]: Following up on https://github.com/bitcoin/bitcoin/pull/13743#discussion_r224988537, I think it would be good to add a comment saying that this has to a separate global instead of a member of MainSignalsInstance, because RegisterWithMempoolSignals is currently called before RegisterBackgroundSignalScheduler, so MainSignalsInstance hasn't been created yet.

Init code is more messy than it needs to be and I think saying what current dependencies are will make it easier to clean up later.

**Review by ryanofsky — APPROVED** [review]
utACK 52532ee43720520696d452424daad213b50aef1f. Changes since last review are adding scripted diff, removing one global, and switching to scoped signal connections.

> Apparently I can't use `${i}` as the script's variable.

It looks like this happens because [commit-script-check.sh](https://github.com/bitcoin/bitcoin/blob/master/test/lint/commit-script-check.sh) uses `$i` and calls `eval "$SCRIPT"` without a subshell. It would be nice to fix this to prevent confusion and ensure the diff checking is correct.

**Review by ryanofsky — APPROVED** [review]
utACK cb53b825c26af6e628ba88d72b2000e75bedbbc6. Only change is last review is new comment.

**Comment by sipa** [comment]
utACK cb53b825c26af6e628ba88d72b2000e75bedbbc6

**Comment by ryanofsky** [comment]
Can this be merged, or does it need more review? I see review from @theuni was requested in github, but I don't know if it's necessary given the other reviews since this change is pretty straightforward.

**Comment by fanquake** [comment]
utACK cb53b82

**Comment by practicalswift** [comment]
utACK cb53b825c26af6e628ba88d72b2000e75bedbbc6

**Pull request merged**
